api/v1alpha4/openstackcluster_webhook.go

```diff
@@ -17,16 +17,64 @@ limitations under the License.
 package v1alpha4
 
 import (
+	"reflect"
+
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/util/validation/field"
 	"sigs.k8s.io/controller-runtime/pkg/builder"
 	logf "sigs.k8s.io/controller-runtime/pkg/log"
 	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/webhook"
 )
 
 // log is for logging in this package.
 var _ = logf.Log.WithName("openstackcluster-resource")
 
+// OpenStackClusterSpecIdentityRefKindImmutableMsg is immutable.
+const OpenStackClusterSpecIdentityRefKindImmutableMsg = "OpenStackCluster spec.IdentityRef.kind field is immutable. Please create a new resource instead."
+
 func (r *OpenStackCluster) SetupWebhookWithManager(mgr manager.Manager) error {
 	return builder.WebhookManagedBy(mgr).
 		For(r).
 		Complete()
 }
+
+// +kubebuilder:webhook:verbs=create;update,path=/validate-infrastructure-cluster-x-k8s-io-v1alpha4-openstackcluster,mutating=false,failurePolicy=fail,matchPolicy=Equivalent,groups=infrastructure.cluster.x-k8s.io,resources=openstackcluster,versions=v1alpha4,name=validation.openstackcluster.infrastructure.x-k8s.io,sideEffects=None,admissionReviewVersions=v1beta1
+
+var _ webhook.Validator = &OpenStackCluster{}
+
+// ValidateCreate implements webhook.Validator so a webhook will be registered for the type.
+func (r *OpenStackCluster) ValidateCreate() error {
+	var allErrs field.ErrorList
+
+	if r.Spec.IdentityRef != nil {
```

Comment:
I think we have to validate that IdentityRef != nil and return an error if it is nil (otherwise we'll get nil pointer dereferences in the controller)

Reply:
Or should we just create the `OpenStackIdentityReference`? As `IdentityRef` is optional, I would assume to not return an error here (`CloudsSecret` was also optional).

Reply:
e.g.

```go
	} else {
		r.Spec.IdentityRef = new(OpenStackIdentityReference)
	}
```

Reply:
Hm no sorry your current version is correct. I didn't remember we pick the credentials up from the environment if there is referenced cloud secret. So code is good as is.